### ISSUE TITLE
Some bugs in runner.sh

### ISSUE THREAD

**Issue by dhouck** [ssafar/effulgence2epub#1]
There seem to be a few problems in `runner.sh`, and I’m not quite sure how all of them should be fixed. What follows is a dump of what I think each stage is and should be doing, folllowed by a summary and an attached patch that (I think) will make it working correctly. Since I don’t fully understand the stages and what’s going on, though, I won’t make it a pull request.

When I use `runner.sh`, it completes firstflat_download correctly, but the next stage (firstflat_parse) fails to run correctly. I believe it should be running `src/extract_firstflat_info.py` , not `src/extract_all.py` (and `extract_firstflat_info.py` appears to need refactoring; it seems to assume that both `web_cache` and `global_lists` are the same directory).  Once I ran that manually, `src/extract_all` still seems to have been the wrong choice (it seems to assume all the pages are downloaded, not just the first ones). I think it should be run after `all_flat_download`, after that’s corrected (see below).

There is also a problem with the stage `all_flat_download`. It seems to be the second half of `images_get` and not downloading the flat pages at all, but in order to run correctly it needs the first half. Moreover, as there are images that aren’t on the first page, `images_get` should probably run after a correct `all_flat_download`. I think `all_flat_download` is meant to be run on the `all_chapters_list` file generated by `extract_firstflat_info.py`.

Finally, at some point (I think before images_get, but I’m not certain), I think there needs to be an `all_flat_parse` stage which does (almost) what `firstflat_parse` does now (`src/extract_all.py` in parallel). Apparently not all distributions are shipping the latest version of GNU Parallel, though, and the command line you use doesn’t work on old versions unless you add a `--gnu` switch (which is o no-op on newer versions).

TL;DR I think, but am not certain, that `runner.sh` needs to have:
- The `firstflat_parse` stage renamed to `all_flat_parse`
- This stage to have the `--gnu` switch added to `parallel`
- A new `firstflat_parse` stage to run `src/extract_firstflat_info.py` and move the output
- The `all_flat_download` stage to be run on `all_chapters_list.txt` instead of `all_the_image_urls.txt`
- The documentation needs to be updated to reflect the above, and the `toc_xhtmlize` stage (I didn’t mention that part above)
- Itself replaced by a `Makefile`, `tupfile`, set of `redo` files (which I hadn’t heard of until yesterday), or similar. That’s a different issue, though (and one that I’m probably not going to bother opening).

Patch: 

``` diff
diff --git a/runner.sh b/runner.sh
index 11355fe..224a524 100755
--- a/runner.sh
+++ b/runner.sh
@@ -13,7 +13,8 @@ cd `dirname $0`
 case $1 in
     "") cat <<EOF
 Please choose between the following operations: toc_download, toc_parse,
-firstflat_download, firstflat_parse, all_flat_download, images_get, gen_ebook.
+firstflat_download, firstflat_parse, all_flat_download, all_flat_parse,
+images_get, toc_xhtmlize, gen_ebook.

 (Typically, this is a reasonable ordering for downloading everything.)
 EOF
@@ -41,8 +42,21 @@ EOF
             wget -i ../global_lists/first_flat_list.txt --no-clobber -w $WAIT_TIME \
             --force-directories)
         ;;
-
+    
     firstflat_parse)
+        (cd web_cache; \
+            cat ../global_lists/chapters.pbtxt | \
+            python ../src/extract_firstflat_info.py; \
+            mv all_chapters_list.txt chapters_with_intros.pbtxt ../global_lists)
+        ;;
+
+    all_flat_download)
+        (cd web_cache; \
+            wget -i ../global_lists/all_chapters_list.txt --no-clobber \
+            -w $WAIT_TIME --force-directories)
+        ;;
+
+    all_flat_parse)
         # Here we use GNU Parallel to launch the processing nicely. We could
         # pipe the entire thing to the parser instead but it wouldn't be as fast
         # / cool.
@@ -51,17 +65,11 @@ EOF
         # at 'chapter {'), and applies extract_all.py on all the chapters
         # independently.
         cat global_lists/chapters_with_intros.pbtxt | \
-            parallel -N1 --pipe --recstart 'chapter {' python src/extract_all.py 
+            parallel --gnu -N1 --pipe --recstart 'chapter {' python src/extract_all.py 
         # We need -N1, otherwise the split wouldn't be potentially happening at
         # every line.
         ;;

-    all_flat_download)
-        (cd web_cache; \
-            wget -i ../global_lists/all_the_image_urls.txt --no-clobber \
-            -w $WAIT_TIME --force-directories)
-        ;;
-
     images_get)
         # Extract all the icon URLs from the chapters. We get the relevant
         # lines, throw away everything but the URL, and download everything
```

**Comment by dhouck** [comment]
Looking at the commit history, it appears you have every chunk I do practically character-for-character, except the documentation one at the top of `runner.sh`

**Comment by ssafar** [comment]
... and the --gnu one; thanks for all of them! (Soon to be fixed!)

As for the build system, I've been thinking about "maybe use redo?" too, but haven't really gotten farther. I've used redo before, not many Makefiles and only cursorily heard of the existence of tup; if you have a bit more experience, which one do you think would be best for this thing?

**Comment by dhouck** [comment]
I only heard of redo when I checked what other repos you had. Tup is similar; it processes the dependencies in “reverse” order. I think it’s easier to use, but that may just be because I’ve used it before and haven’t used redo before. All three systems, and probably several others, would allow good parallelization and easier building.

Tup is probably the best to use of the ones I know, because it has good configuration management, proper dependency handling (including complex dependencies that I know little about but think might be helpful here), and good cross-platform support. On the other hand, Make has the advantage of being on practically all the machines that could also run the shell scripts, not just supporting those machines.

**Comment by ssafar** [comment]
Added docs & "--gnu" in the commit above. (I guess we do include everything now, do we?)

**Comment by dhouck** [comment]
I do believe so. The doc works, and my computer is able to run the things involving `parallel`.
